File: src/effects/NoiseGate.cpp
#include <atomic>
#include <cmath>

#include "effect.h"
#include "registry.h"

// peak-follower gate: opens fast when the signal crosses the threshold and closes
// over the release time. the applied gain is smoothed so it does not click.
class NoiseGate : public Effect {
public:
  NoiseGate() {
    param("threshold", &threshold, 0.0f, 0.3f);
    param("release", &releaseMs, 10.0f, 1000.0f);
  }

  void prepare(double sampleRate, int) override {
    sr_ = (float)sampleRate;
    env_ = 0.0f;
    gain_ = 0.0f;
  }

  void process(float* buffer, int numFrames) override {
    float thr = threshold.load(std::memory_order_relaxed);
    float rel = releaseMs.load(std::memory_order_relaxed);
    float relCoef = std::exp(-1.0f / ((rel * 0.001f) * sr_));
    float atkCoef = std::exp(-1.0f / (0.002f * sr_));  // 2 ms open ramp
    for (int i = 0; i < numFrames; i++) {
      float a = std::fabs(buffer[i]);
      env_ = a > env_ ? a : env_ * relCoef;  // instant attack, decaying release
      float target = env_ >= thr ? 1.0f : 0.0f;
      float c = target > gain_ ? atkCoef : relCoef;
      gain_ = target + (gain_ - target) * c;
      buffer[i] *= gain_;
    }
  }
  const char* name() const override { return "noise gate"; }

private:
  std::atomic<float> threshold{0.02f};
  std::atomic<float> releaseMs{200.0f};
  float sr_ = 48000.0f;
  float env_ = 0.0f;
  float gain_ = 0.0f;
};

REGISTER_EFFECT(NoiseGate, "noise gate")

File: effects/noisegate.lua
-- peak-follower gate: opens fast when the signal crosses the threshold and
-- closes over the release time. the applied gain is smoothed so it does not click.
name = "noise gate"
params = {
  { name = "threshold", min = 0,  max = 0.3,  default = 0.02 },
  { name = "release",   min = 10, max = 1000, default = 200 },
}

local exp, abs = math.exp, math.abs

local sr, env, gain = 48000, 0, 0

function prepare(sampleRate, maxBlock)
  sr = sampleRate
  env = 0
  gain = 0
end

function process(buf, n, p)
  local relCoef = exp(-1 / (p.release * 0.001 * sr))
  local atkCoef = exp(-1 / (0.002 * sr))
  local thr = p.threshold
  for i = 0, n - 1 do
    local a = abs(buf[i])
    if a > env then env = a else env = env * relCoef end
    local target = env >= thr and 1 or 0
    local c = target > gain and atkCoef or relCoef
    gain = target + (gain - target) * c
    buf[i] = buf[i] * gain
  end
end
